### ISSUE TITLE
Question regarding DRY in sagas

### ISSUE THREAD

**Issue by msikma** [yelouafi/redux-saga#638]
Redux Saga seems to be precisely what I need in my project, for cleanly doing async work using Redux. However there's something I don't really understand that well, and I can't seem to really figure it out using the docs.

In my app, let's say I've got a button which fires `ASYNC_ACTION_START` when clicked. I've implemented the first example to connect this to a saga:

```js
function* doAsyncActionOne(action) {
   try {
      const result = yield call(Api.doThingOne, action.payload.someArgument);
      yield put({type: "ASYNC_ACTION_SUCCEEDED", result});
   } catch (e) {
      yield put({type: "ASYNC_ACTION_FAILED", message: e.message});
   }
}
function* mySaga() {
  yield* takeLatest("ASYNC_ACTION_START", doAsyncActionOne);
}
export default function* root() {
  yield takeLatest("ASYNC_ACTION_START", doAsyncActionOne);
}
[...]
sagaMW.run(root);
```

This works perfectly fine. Now I add another button, which does almost the same thing, except it calls a different API function with slightly different arguments.

So I try to deduplicate this, here's my naive implementation:

```js
import { takeLatest } from 'redux-saga';
import { call, put } from 'redux-saga/effects';

import { Api.doThingOne, Api.doThingTwo } from '../api';

function* apiCall(fn, successType, failType, ...args) {
  // The original action is always passed as the last argument.
  const action = args.pop();
  let callArgs = [];

  // Action 2 has email/password arguments.
  if (action.type === 'ASYNC_ACTION_2_START') {
    callArgs = [action.email, action.password];
  }
  
  try {
    const payload = yield call(fn, ...callArgs);
    yield put({ type: successType, payload });
  }
  catch (e) {
    yield put({ type: failType, payload: e });
  }
}

export default function* root() {
  yield takeLatest('ASYNC_ACTION_START', apiCall, Api.doThingOne, 'ASYNC_ACTION_SUCCEEDED', 'ASYNC_ACTION_FAILED');
  yield takeLatest('ASYNC_ACTION_2_START', apiCall, Api.doThingTwo, 'ASYNC_ACTION_2_SUCCEEDED', 'ASYNC_ACTION_2_FAILED');
}
```

There's one thing that bothers me about this in particular, which is that I don't have access to the action in `root()` so I can't just send `action.email` and `action.password` directly to `apiCall()`. It does work though. Originally I simply had two identical functions like the one above, except with that one difference of adding some extra arguments to the call.

Am I doing things totally wrong, or awkwardly? Is there a recommended way to do this?

edit: now the issue is complete, sorry for accidentally submitting a half written version first.

**Comment by Andarist** [comment]
Well, its not rly a redux-saga problem, but the one much broader - programming and DRY in general. The lib cannot give you here a golden way of doing such things.

If you want to deduplicate such things I would go probably with making a `spec` (object) argument instead of multiple arguments. I prefer this way with more than 3 arguments as its just easier to remember what the function takes as arguments in its `spec` than to remember ordering of those.

Then you can easily add 1 more option to the `spec`, smth like `actionToArgs` which would be a function taking an action and returning your `callArgs`.


It could become smth like this:
```javascript
const constant = (item) => () => item

function* apiCall(fn, {
  successType,
  failType,
  actionToArgs = constant([]),
}, action) {
  try {
    const payload = yield call(fn, ...actionToArgs(callArgs));
    yield put({ type: successType, payload });
  }
  catch (e) {
    yield put({ type: failType, payload: e });
  }
}

export default function* root() {
  yield takeLatest('ASYNC_ACTION_START', apiCall, Api.doThingOne, {
    successType: 'ASYNC_ACTION_SUCCEEDED',
    failType: 'ASYNC_ACTION_FAILED'
  });
  yield takeLatest('ASYNC_ACTION_2_START', apiCall, Api.doThingTwo, {
    successType: 'ASYNC_ACTION_2_SUCCEEDED',
    failType: 'ASYNC_ACTION_2_FAILED',
    actionToArgs: ({ email, password  }) => [email, password]
  });
}
```

**Comment by msikma** [comment]
Thanks a lot, sorry for taking some time to reply to this. You've given me something good to think about, and I did a couple of different tests to try and work out what would be best for my case. In the end I also decided that I wanted to call an action rather than simply `put()` the success action. So I decided to split that particular saga off into its own function.

Still, this is very helpful, it's good to get a perspective from someone knowledgable on how to work with sagas. Thanks a lot for your thoughts 👍
